src/main/scala/calc/Typechecker.scala

```diff
@@ -0,0 +1,138 @@
+package calc
+
+import org.scalajs.core.ir.{Types => irtpe, Position}
+
+/** A simple typechecker.
+ *
+ * Checks whether the types match or are otherwise appropriate (e.g. in the if's condition).
+ * Annotates AST with types, so that they can be used during compilation.
+ *
+  * TODO group errors and fail as late as possible, possibly without an exception (requires some type-work)
+ */
+
+object Typechecker {
+
+  class TypecheckerException(msg: String) extends Exception(msg)
+
+  type TypeScope = Map[String, TreeType]
+
+  implicit def getScalaJSType(tp: TreeType): irtpe.Type = tp match {
+    case DoubleType      => irtpe.DoubleType
+    case FunctionType(_) => irtpe.AnyType
+    case NoType          => irtpe.NoType
+  }
+
+  def getType(tree: Tree, scope: TypeScope)
+             (implicit pos: Position): TypedTree =  tree match {
```

Comment:
The `implicit pos` is suspicious, here, as any necessary position should derived from `tree.pos` at this point.